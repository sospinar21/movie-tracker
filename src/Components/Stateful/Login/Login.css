.login-container {
  width: 100vw;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column; }

.login {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  background-color: #353434;
  height: 320px;
  width: 600px;
  border-radius: 10px; }

.login-input {
  height: 38px;
  width: 300px;
  margin: 20px 8px;
  border: solid 2px #26a566;
  border-radius: 5px;
  padding-left: 10px;
  font-size: 20px; }

.email-input {
  margin-left: 44px; }

.login-input::placeholder {
  font-size: 20px; }

label {
  font-size: 20px;
  color: white; }

.login-button {
  margin: 20px 0;
  width: 420px;
  height: 38px;
  font-size: 18px;
  background-color: #26a566;
  color: white;
  border-radius: 5px;
  transition: 0.3s;
  box-shadow: 0 10px 18px #494444;
  border: none;
  outline: none; }

h2 {
  color: #353434; }

.login-button:hover {
  cursor: pointer;
  /* background-color: rgb(30, 90, 202); */
  box-shadow: 0 6px 18px #777; }
