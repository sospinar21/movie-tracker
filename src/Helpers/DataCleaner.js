class DataCleaner {

  cleanMovies = (data) => {
    const cleanData = data.map(movie => ({
      title: movie.original_title,
      image: movie.poster_path,
      id: movie.id,
      releaseDate: movie.release_date,
      voteAverage: movie.vote_average,
      overview: movie.overview
    }))

    return cleanData;
  }

}

export default DataCleaner;