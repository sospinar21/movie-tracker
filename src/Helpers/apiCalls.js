import { apikey } from '../apikey';
import DataCleaner from './DataCleaner';

export const fetchMovieData = async (user) => {
  try {
    const dataCleaner = new DataCleaner();
    const url = `https://api.themoviedb.org/3/movie/now_playing?api_key=${apikey}&language=en-US&page=1`;
    const response = await fetch(url);
    const data = await response.json();
    const cleanData = await dataCleaner.cleanMovies(data.results, user);
    
    return cleanData;
  } catch (error) {
    throw new Error('Failed to fetch');
  }
};

export const signIn = async (email, password) => {
  try {
    const url = '/api/users';
    const login = {email, password};
    const stringifiedBody = JSON.stringify(login);
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-type': 'application/json'
      },
      body: stringifiedBody
    });
    if (response.ok) {
      const user = await response.json();
      return user;
    } else {
      const data = {data: 'Invalid Username or Password'};
      return data;
    }
  } catch (error) {
    throw error;
  }
};

export const signUp = async (name, email, password) => {
  try {
    const url = '/api/users/new';
    const body = {name, email, password};
    const stringifiedBody = JSON.stringify(body);
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-type': 'application/json'
      },
      body: stringifiedBody
    });
    if (response.ok) {
      const user = await response.json();
      return user;
    } else {
      const data = {data: 'Email is already taken'};
      return data;
    }
  } catch (error) {
    throw error;
  }
};

export const postMovieToFavorites = async (movie, favorites) => {
  try {
    const url = '/api/users/favorites/new';
    const newMovie = JSON.stringify(movie);
    const match = favorites.find(favorite => favorite.favoriteId === movie.movie_id);
    if (!match) {
      const response  = await fetch(url, {
        method: 'POST', 
        body: newMovie,
        headers: new Headers({
          'Content-Type': 'application/json'
        })
      });
    }
  } catch (error) {
    throw error;
  }
};

export const retrieveFavorites = async (userId) => {
  try {
    const cleaner = new DataCleaner();
    const url = `/api/users/${userId}/favorites`;
    const response = await fetch(url);
    if (response.ok) {
      const favorites = await response.json();
      const favId = await cleaner.cleanFavorites(favorites.data);
      return favId;
    }
  } catch (error) {
    throw error;
  }
};

export const deleteFavorite = async (userId, movieId) => {
  try {
    const url = `/api/users/${userId}/favorites/${movieId}`;
    const response = await fetch(url, {
      method: 'DELETE'
    });
  } catch (error) {
    throw error;
  }
};