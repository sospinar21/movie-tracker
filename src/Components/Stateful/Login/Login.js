import React, { Component } from 'react';
import './Login.css';
import { signIn, retrieveFavorites} from './../../../Helpers/apiCalls';
import { login, addFavorites } from '../../../Actions/actions';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';


export class Login extends Component {
  constructor() {
    super();
    this.state = {
      email: '',
      password: '',
      status: ''
    };
  }

  handleSubmit = async (event) => {
    const email = this.state.email.toLowerCase();
    event.preventDefault();
    const user = await signIn(email, this.state.password);
    this.props.login(user.data);
    this.setState({
      email: '',
      password: '',
      status: this.loginAuthorization()
    });
  }
  
  loginAuthorization = () => {
    if (this.props.user.name) {
      this.addFavorites();
      return <Redirect to='/'/>;
    } else {
      return (
        <h2 className='loginh2'> {this.props.user} </h2>
      );
    }
  }

  addFavorites = async () => {
    const userFavorites = await retrieveFavorites(this.props.user.id);
    this.props.addFavorites(userFavorites);
  }

  handleSignin = (event) => {
    const {name, value} = event.target;
    this.setState({
      [name]: value
    });
  }

  render() {
    
    return (
      <div className='login-container'>
        <h2> Login to start favoriting movies! </h2>
        <form onSubmit={this.handleSubmit}
          className='login'>
          <label className='email-label'>Email
            <input type="text" 
              placeholder='*JohnDoe@gmail.com*'
              onChange={this.handleSignin}
              value={this.state.email}
              name="email"
              className='login-input email-input'/>
          </label>
          <label className='password-label'>Password 
            <input type="password" 
              onChange={this.handleSignin}
              value={this.state.password}
              name="password"
              className='login-input'/>
          </label>
          <button className='login-button'>Login</button>
        </form>
        { this.state.status }
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  user: state.user
}); 

export const mapDispatchToProps = (dispatch) => ({
  login: (user) => dispatch(login(user)),
  addFavorites: (movies) => dispatch(addFavorites(movies))
});

Login.propTypes = {
  user: PropTypes.obj,
  login: PropTypes.func,
  addFavorites: PropTypes.func
};

export default connect(mapStateToProps, mapDispatchToProps)(Login);