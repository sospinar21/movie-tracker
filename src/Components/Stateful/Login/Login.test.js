import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import React from 'react';
import { Login, mapStateToProps, mapDispatchToProps } from './Login';
import { signIn } from '../../../Helpers/apiCalls';
import { Redirect } from 'react-router-dom';

describe('login', () => {

  let login;
  let mockProps;
  let event;

  beforeEach(() => {

    mockProps = {
      user: {name: 'Jack'},
      login: jest.fn(),
      addFavorites: jest.fn()
    }

    login = shallow(<Login {...mockProps} />)

  })

  describe('Component', () => {

    it('matches snapshot', () => {

      login = shallow(<Login {...mockProps} />, { disableLifecycleMethods: true });
      expect(login).toMatchSnapshot();

    })
  })

  describe('handleSubmit', () => {

    let event;
    window.fetch = jest.fn()

    beforeEach(() => {

      event = { preventDefault: () => {} };
      window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        status: 200,
        json: () => Promise.resolve({data: {}})
      }));
    })

    it('calls props.login', async () => {

      await login.instance().handleSubmit(event);
      
      expect(mockProps.login).toHaveBeenCalled();
    })

    it('resets the state', async () => {
      
      login.state({
        email: 'sospinar@gmail.com',
        password: 'hello'
      }) 

      await login.instance().handleSubmit(event);
      
      expect(login.state().email).toEqual('');
      expect(login.state().password).toEqual('');      

    })
  })

  describe('loginAuthorization', () => {

    it('displays a message if the user is not logged in', () => {

      mockProps.user = {name: ''}
      
      const authorization = login.instance().loginAuthorization();

      expect(authorization).toEqual(<Redirect push={false} to="/" />);
    })

    it('displays a succesfull message if the user is logged in', () => {

      const mockProps = { 
        user: {name: ''},
        login: jest.fn(),
        addFavorites: jest.fn()
      }

      const expected = {'name': ''}
      login = shallow(<Login {...mockProps} />)

      const authorization = login.instance().loginAuthorization();

      expect(authorization).toEqual(<h2 className="loginh2"> {expected} </h2>);
    })

  })

  describe('addFavorites', () => {
    
    it('calls addFavorites', async () => {

      await login.instance().addFavorites();

      expect(mockProps.addFavorites).toHaveBeenCalled();

    })
  })

  describe('mapStateToProps', () => {
    it('returns an object with the users info', () => {

      const mockState = {
        user: {name: 'Steph'}
      }

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(mockState);
      
    })
  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch with the correct params', () => {

      const mockDispatch = jest.fn();
      const user = {name: 'jack'}
      const mappedProps = mapDispatchToProps(mockDispatch);
      const mockAction = {
        type: 'LOGIN',
        user
      };
      
      mappedProps.login(user);
      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
      
    })
  })
})