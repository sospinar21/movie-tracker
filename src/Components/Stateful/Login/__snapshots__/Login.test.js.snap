// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login Component matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    addFavorites={[Function]}
    login={[Function]}
    user={
        Object {
            "name": "Jack",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
           Login to start favoriting movies! 
</h2>,
        <form
          className="login"
          onSubmit={[Function]}
>
          <label
                    className="email-label"
          >
                    Email
                    <input
                              className="login-input email-input"
                              name="email"
                              onChange={[Function]}
                              placeholder="*JohnDoe@gmail.com*"
                              type="text"
                              value=""
                    />
          </label>
          <label
                    className="password-label"
          >
                    Password
                    <input
                              className="login-input"
                              name="password"
                              onChange={[Function]}
                              type="password"
                              value=""
                    />
          </label>
          <button
                    className="login-button"
          >
                    Login
          </button>
</form>,
        "",
      ],
      "className": "login-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": " Login to start favoriting movies! ",
        },
        "ref": null,
        "rendered": " Login to start favoriting movies! ",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label
              className="email-label"
>
              Email
              <input
                            className="login-input email-input"
                            name="email"
                            onChange={[Function]}
                            placeholder="*JohnDoe@gmail.com*"
                            type="text"
                            value=""
              />
</label>,
            <label
              className="password-label"
>
              Password
              <input
                            className="login-input"
                            name="password"
                            onChange={[Function]}
                            type="password"
                            value=""
              />
</label>,
            <button
              className="login-button"
>
              Login
</button>,
          ],
          "className": "login",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Email",
                <input
                  className="login-input email-input"
                  name="email"
                  onChange={[Function]}
                  placeholder="*JohnDoe@gmail.com*"
                  type="text"
                  value=""
/>,
              ],
              "className": "email-label",
            },
            "ref": null,
            "rendered": Array [
              "Email",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "login-input email-input",
                  "name": "email",
                  "onChange": [Function],
                  "placeholder": "*JohnDoe@gmail.com*",
                  "type": "text",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
            ],
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Password",
                <input
                  className="login-input"
                  name="password"
                  onChange={[Function]}
                  type="password"
                  value=""
/>,
              ],
              "className": "password-label",
            },
            "ref": null,
            "rendered": Array [
              "Password",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "login-input",
                  "name": "password",
                  "onChange": [Function],
                  "type": "password",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
            ],
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Login",
              "className": "login-button",
            },
            "ref": null,
            "rendered": "Login",
            "type": "button",
          },
        ],
        "type": "form",
      },
      "",
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
             Login to start favoriting movies! 
</h2>,
          <form
            className="login"
            onSubmit={[Function]}
>
            <label
                        className="email-label"
            >
                        Email
                        <input
                                    className="login-input email-input"
                                    name="email"
                                    onChange={[Function]}
                                    placeholder="*JohnDoe@gmail.com*"
                                    type="text"
                                    value=""
                        />
            </label>
            <label
                        className="password-label"
            >
                        Password
                        <input
                                    className="login-input"
                                    name="password"
                                    onChange={[Function]}
                                    type="password"
                                    value=""
                        />
            </label>
            <button
                        className="login-button"
            >
                        Login
            </button>
</form>,
          "",
        ],
        "className": "login-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": " Login to start favoriting movies! ",
          },
          "ref": null,
          "rendered": " Login to start favoriting movies! ",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label
                className="email-label"
>
                Email
                <input
                                className="login-input email-input"
                                name="email"
                                onChange={[Function]}
                                placeholder="*JohnDoe@gmail.com*"
                                type="text"
                                value=""
                />
</label>,
              <label
                className="password-label"
>
                Password
                <input
                                className="login-input"
                                name="password"
                                onChange={[Function]}
                                type="password"
                                value=""
                />
</label>,
              <button
                className="login-button"
>
                Login
</button>,
            ],
            "className": "login",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Email",
                  <input
                    className="login-input email-input"
                    name="email"
                    onChange={[Function]}
                    placeholder="*JohnDoe@gmail.com*"
                    type="text"
                    value=""
/>,
                ],
                "className": "email-label",
              },
              "ref": null,
              "rendered": Array [
                "Email",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "login-input email-input",
                    "name": "email",
                    "onChange": [Function],
                    "placeholder": "*JohnDoe@gmail.com*",
                    "type": "text",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
              ],
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Password",
                  <input
                    className="login-input"
                    name="password"
                    onChange={[Function]}
                    type="password"
                    value=""
/>,
                ],
                "className": "password-label",
              },
              "ref": null,
              "rendered": Array [
                "Password",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "login-input",
                    "name": "password",
                    "onChange": [Function],
                    "type": "password",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
              ],
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Login",
                "className": "login-button",
              },
              "ref": null,
              "rendered": "Login",
              "type": "button",
            },
          ],
          "type": "form",
        },
        "",
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
